### PR TITLE
Cleanup Hda client from root-core

### PR DESCRIPTION
**Pull request by peter-rustler-basyskom**

### CODE REVIEW

**Comment by basyskom-meerkoetter** [comment]
@Root-Core we are working on finishing the (longstanding) HDA client PR of yours. Please let us know if you have the need to coordinate.

File: tests/client/historical_read_test_data.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ */
+
+/* Data in this header is gathered from OPC Unified Architecture, Part 11, Release
+1.03 Page 5-6 from OPC Foundation */
+
+#ifndef UA_HISTORICAL_READ_TEST_DATA_H_
+#define UA_HISTORICAL_READ_TEST_DATA_H_
+
+#include "ua_types.h"
+#include <limits.h>
```
Comment (basyskom-jvoe) [inline tests/client/historical_read_test_data.h]: Is this still required?

File: examples/client_historical.c [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    printf("\nRead historical callback:\n");
+    printf("\tValue count:\t%u\n", (UA_UInt32)_data->dataValuesSize);
+    printf("\tHas more data:\t%d\n\n", moreDataAvailable);
+
+    /* Iterate over all values */
+    for (UA_UInt32 i = 0; i < _data->dataValuesSize; ++i)
+    {
+        UA_DataValue val = _data->dataValues[i];
+        
+        /* If there is no value, we are out of bounds or something bad hapened */
```
[the remark this answers is not in the document]
Comment (peter-rustler-basyskom) [reply examples/client_historical.c]: done

File: include/ua_client_highlevel.h [hunk 3]
```diff
@@ -210,6 +210,44 @@ UA_Client_readUserExecutableAttribute(UA_Client *client, const UA_NodeId nodeId,
...
+                                 const UA_NodeId *nodeId,
+                                 UA_Boolean moreDataAvailable,
+                                 const UA_ExtensionObject *data, void *callbackContext);
+
+#ifdef UA_ENABLE_EXPERIMENTAL_HISTORIZING
+UA_StatusCode UA_EXPORT
+UA_Client_HistoryRead_events(UA_Client *client, const UA_NodeId nodeId,
+                                const UA_HistoricalIteratorCallback callback,
+                                const UA_DateTime startTime, const UA_DateTime endTime,
+                                UA_String indexRange, const UA_EventFilter filter, const UA_UInt32 numValuesPerNode,
```
[the remark this answers is not in the document]
Comment (peter-rustler-basyskom) [reply include/ua_client_highlevel.h]: done

File: include/ua_client_highlevel.h [hunk 4]
```diff
@@ -210,6 +210,44 @@ UA_Client_readUserExecutableAttribute(UA_Client *client, const UA_NodeId nodeId,
...
+#ifdef UA_ENABLE_HISTORIZING
+typedef UA_Boolean
+(*UA_HistoricalIteratorCallback)(UA_Client *client,
+                                 const UA_NodeId *nodeId,
+                                 UA_Boolean moreDataAvailable,
+                                 const UA_ExtensionObject *data, void *callbackContext);
+
+#ifdef UA_ENABLE_EXPERIMENTAL_HISTORIZING
+UA_StatusCode UA_EXPORT
+UA_Client_HistoryRead_events(UA_Client *client, const UA_NodeId nodeId,
```
[the remark this answers is not in the document]
Comment (peter-rustler-basyskom) [reply include/ua_client_highlevel.h]: done

File: tests/client/historical_read_test_data.h (hunk 1 above)
Comment (peter-rustler-basyskom) [reply tests/client/historical_read_test_data.h]: done

**Comment by Pro** [comment]
Cool, thanks for updating and finalizing this feature.

@peter-rustler-basyskom can you squash the commits from @Root-Core into one (also squash mine), and yours into one, so that at the end we have two commits since it's all the same feature? Or if necessary split it in less commits to keep the commit history clean.

**Comment by basyskom-meerkoetter** [comment]
> Cool, thanks for updating and finalizing this feature.
> 
> @peter-rustler-basyskom can you squash the commits from @Root-Core into one (also squash mine), and yours into one, so that at the end we have two commits since it's all the same feature? Or if necessary split it in less commits to keep the commit history clean.

@Pro @peter-rustler-basyskom will be back on the 7th of January. Just to let you know

File: examples/client_historical.c [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+#endif
+#ifndef _DEFAULT_SOURCE
+# define _DEFAULT_SOURCE
+#endif
+/* On older systems we need to define _BSD_SOURCE.
+ * _DEFAULT_SOURCE is an alias for that. */
+#ifndef _BSD_SOURCE
+# define _BSD_SOURCE
+#endif
+
```
Comment (Pro) [inline examples/client_historical.c]: These defines should not be necessary, it's already in open62541.h

File: examples/client_historical.c [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+#include <stdio.h>
+#include "open62541.h"
+
+#ifdef _WIN32
+# include <windows.h>
+# define UA_sleep_ms(X) Sleep(X)
+#else
+# include <unistd.h>
+# define UA_sleep_ms(X) usleep(X * 1000)
+#endif
```
Comment (Pro) [inline examples/client_historical.c]: UA_sleep_ms should also already be defined inside the stack.
E.g. see here: https://github.com/open62541/open62541/blob/master/examples/client_connect_loop.c#L69

File: include/ua_client_highlevel.h [hunk 7]
```diff
@@ -210,6 +212,44 @@ UA_Client_readUserExecutableAttribute(UA_Client *client, const UA_NodeId nodeId,
...
+#ifdef UA_ENABLE_HISTORIZING
+typedef UA_Boolean
+(*UA_HistoricalIteratorCallback)(UA_Client *client,
+                                 const UA_NodeId *nodeId,
+                                 UA_Boolean moreDataAvailable,
+                                 const UA_ExtensionObject *data, void *callbackContext);
+
+#ifdef UA_ENABLE_EXPERIMENTAL_HISTORIZING
+UA_StatusCode UA_EXPORT
+UA_Client_HistoryRead_events(UA_Client *client, const UA_NodeId *nodeId,
```
Comment (Pro) [inline include/ua_client_highlevel.h]: We normally use a lower case letter after the underscore. I.e. this should be `UA_Client_historyRead_events` to be consistent.

Also for the other definitions below.

Comment (jpfr) [reply include/ua_client_highlevel.h]: Nope. In the client we have like "modules" that are CamlCase starting upper-case.
The first verb starts lowercase...

For example `UA_Client_MonitoredItems_modify`

Comment (Pro) [reply include/ua_client_highlevel.h]: Ah, right, sorry, then that's correct :+1:

**Comment by Pro** [comment]
Thank you very much!!

**Pull request merged**